README.md
# Algolia Search API Client for Swift

**&lt;Welcome Objective-C developers&gt;**

In July 2015, we release a new version of our Swift client able to work with Swift and Objective-C.

If you were using our Objective-C client, [read the migration guide](https://github.com/algolia/algoliasearch-client-swift/wiki/Migration-guide-from-Objective-C-to-Swift-API-Client).

The Objective-C API Client is still supported and can be found [here](https://github.com/algolia/algoliasearch-client-objc).

**&lt;/Welcome Objective-C developers&gt;**




[Algolia Search](http://www.algolia.com) is a hosted full-text, numerical, and faceted search engine capable of delivering realtime results from the first keystroke.

Our Swift client lets you easily use the [Algolia Search API](https://www.algolia.com/doc/rest_api) from your backend. It wraps the [Algolia Search REST API](http://www.algolia.com/doc/rest_api).




[![Build Status](https://travis-ci.org/algolia/algoliasearch-client-swift.svg?branch=master)](https://travis-ci.org/algolia/algoliasearch-client-swift)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![CocoaPods](https://img.shields.io/cocoapods/v/AlgoliaSearch-Client-Swift.svg)]()
[![CocoaPods](https://img.shields.io/cocoapods/l/AlgoliaSearch-Client-Swift.svg)]()
[![](http://img.shields.io/badge/OS%20X-10.9%2B-lightgrey.svg)]()
[![](http://img.shields.io/badge/iOS-7.0%2B-lightgrey.svg)]()




Table of Contents
=================
**Getting Started**

1. [Setup](#setup)
1. [Quick Start](#quick-start)
1. [Online documentation](#documentation)
1. [Tutorials](#tutorials)


**Commands Reference**

1. [Add a new object](#add-a-new-object-to-the-index)
1. [Update an object](#update-an-existing-object-in-the-index)
1. [Search](#search)
1. [Search cache](#search-cache)
1. [Multiple queries](#multiple-queries)
1. [Get an object](#get-an-object)
1. [Delete an object](#delete-an-object)
1. [Delete by query](#delete-by-query)
1. [Index settings](#index-settings)
1. [List indices](#list-indices)
1. [Delete an index](#delete-an-index)
1. [Clear an index](#clear-an-index)
1. [Wait indexing](#wait-indexing)
1. [Batch writes](#batch-writes)
1. [Security / User API Keys](#security--user-api-keys)
1. [Copy or rename an index](#copy-or-rename-an-index)
1. [Backup / Retrieve all index content](#backup--retrieve-of-all-index-content)
1. [Logs](#logs)




Setup
-------------
To setup your project, follow these steps:



 1. [Download and add sources](https://github.com/algolia/algoliasearch-client-swift/archive/master.zip) to your project or use [Carthage](https://github.com/Carthage/Carthage) by adding `github "algolia/algoliasearch-client-swift"` in your Cartfile or use cocoapods by adding `pod 'AlgoliaSearch-Client-Swift', '~> 2.0'` in your Podfile.
 2. Add the `import AlgoliaSearch` call to your project
 3. Initialize the client with your ApplicationID and API-Key. You can find all of them on [your Algolia account](http://www.algolia.com/users/edit).

```swift
let client = AlgoliaSearch.Client(appID: "YourApplicationID", apiKey: "YourAPIKey")
```




Quick Start
-------------


In 30 seconds, this quick start tutorial will show you how to index and search objects.

Without any prior configuration, you can start indexing [500 contacts](https://github.com/algolia/algoliasearch-client-csharp/blob/master/contacts.json) in the ```contacts``` index using the following code:
```swift
// Load content file
let jsonPath = NSBundle.mainBundle().pathForResource("contacts", ofType: "json")
let jsonData = NSData.dataWithContentsOfMappedFile("jsonPath")
let dict = NScontentSerialization.contentObjectWithData(jsonData, options: 0, error: nil)
// Load all objects of json file in an index named "contacts"
index.addObjects(dict["objects"], block: nil)
```

You can now search for contacts using firstname, lastname, company, etc. (even with typos):
```swift
// search by firstname
index.search(Query(query: "jimmie"), block: { (content, error) -> Void in
	if error == nil {
		println("Result: \(content)")
	}
})
// search a firstname with typo
index.search(Query(query: "jimie"), block: { (content, error) -> Void in
	if error == nil {
		println("Result: \(content)")
	}
})
// search for a company
index.search(Query(query: "california paint"), block: { (content, error) -> Void in
	if error == nil {
		println("Result: \(content)")
	}
})
// search for a firstname & company
index.search(Query(query: "jimmie paint"), block: { (content, error) -> Void in
	if error == nil {
		println("Result: \(content)")
	}
})
```

Settings can be customized to tune the search behavior. For example, you can add a custom sort by number of followers to the already great built-in relevance:
```swift
let customRanking = ["desc(followers)"]
let settings = ["customRanking": customRanking]
index.setSettings(settings, block: { (content, error) -> Void in
	if let error = error {
		println("Error when applying settings: \(error)")
	}
})
```

You can also configure the list of attributes you want to index by order of importance (first = most important):
```swift
let customRanking = ["lastname", "firstname", "company", "email", "city", "address"]
let settings = ["attributesToIndex": customRanking]
index.setSettings(settings, block: { (content, error) -> Void in
	if let error = error {
		println("Error when applying settings: \(error)")
	}
})
```

Since the engine is designed to suggest results as you type, you'll generally search by prefix. In this case the order of attributes is very important to decide which hit is the best:
```swift
index.search(Query(query: "or"), block: { (content, error) -> Void in
	if error == nil {
		println("Result: \(content)")
	}
})

index.search(Query(query: "jim"), block: { (content, error) -> Void in
	if error == nil {
		println("Result: \(content)")
	}
})
```





Documentation
================
Check our [online documentation](http://www.algolia.com/doc/guides/swift):
 * [Initial Import](http://www.algolia.com/doc/guides/swift#InitialImport)
 * [Ranking &amp; Relevance](http://www.algolia.com/doc/guides/swift#RankingRelevance)
 * [Indexing](http://www.algolia.com/doc/guides/swift#Indexing)
 * [Search](http://www.algolia.com/doc/guides/swift#Search)
 * [Sorting](http://www.algolia.com/doc/guides/swift#Sorting)
 * [Filtering](http://www.algolia.com/doc/guides/swift#Filtering)
 * [Faceting](http://www.algolia.com/doc/guides/swift#Faceting)
 * [Geo-Search](http://www.algolia.com/doc/guides/swift#Geo-Search)
 * [Security](http://www.algolia.com/doc/guides/swift#Security)
 * [REST API](http://www.algolia.com/doc/rest)

Tutorials
================

Check out our [tutorials](http://www.algolia.com/doc/tutorials):
 * [Search bar with autocomplete menu](http://www.algolia.com/doc/tutorials/auto-complete)
 * [Search bar with multi category autocomplete menu](http://www.algolia.com/doc/tutorials/multi-auto-complete)
 * [Instant search result pages](http://www.algolia.com/doc/tutorials/instant-search)




Commands Reference
==================



Add a new object to the Index
-------------

Each entry in an index has a unique identifier called `objectID`. There are two ways to add en entry to the index:

 1. Using automatic `objectID` assignment. You will be able to access it in the answer.
 2. Supplying your own `objectID`.

You don't need to explicitly create an index, it will be automatically created the first time you add an object.
Objects are schema less so you don't need any configuration to start indexing. If you wish to configure things, the settings section provides details about advanced settings.

Example with automatic `objectID` assignment:

```swift
let newObject = ["firstname": "Jimmie", "lastname": "Barninger"]
index.addObject(object, block: { (content, error) -> Void in
	if error == nil {
		let objectID = content!["objectID"] as String
		println("Object ID: \(objectID)")
	}
)
```

Example with manual `objectID` assignment:

```swift
let newObject = ["firstname": "Jimmie", "lastname": "Barninger"]
index.addObject(object, withID: "myID", block: { (content, error) -> Void in
	if error == nil {
		let objectID = content!["objectID"] as String
		println("Object ID: \(objectID)")
	}
)
```

Update an existing object in the Index
-------------

You have three options when updating an existing object:

 1. Replace all its attributes.
 2. Replace only some attributes.
 3. Apply an operation to some attributes.

Example on how to replace all attributes of an existing object:

```swift
let newObject = [
	"firstname": "Jimmie",
	"lastname": "Barninger",
	"city": "New York",
	"objectID": "myID"
]
index.saveObject(newObject, block: nil)
```

You have many ways to update an object's attributes:

 1. Set the attribute value
 2. Add a string or number element to an array
 3. Remove an element from an array
 4. Add a string or number element to an array if it doesn't exist
 5. Increment an attribute
 6. Decrement an attribute

Example to update only the city attribute of an existing object:

```swift
let partialObject = ["city": "San Francisco"]
index.partialUpdateObject(partialObject, objectID: "myID", block: nil)
```

Example to add a tag:

```swift
let operation = [
	"value": "MyTag",
	"_operation": "Add"
]
let partialObject = ["_tags": operation]
index.partialUpdateObject(partialObject, objectID: "myID", block: nil)
```

Example to remove a tag:

```swift
let operation = [
	"value": "MyTag",
	"_operation": "Remove"
]
let partialObject = ["_tags": operation]
index.partialUpdateObject(partialObject, objectID: "myID", block: nil)
```

Example to add a tag if it doesn't exist:

```swift
let operation = [
	"value": "MyTag",
	"_operation": "AddUnique"
]
let partialObject = ["_tags": operation]
index.partialUpdateObject(partialObject, objectID: "myID", block: nil)
```

Example to increment a numeric value:

```swift
let operation = [
	"value": 42,
	"_operation": "Increment"
]
let partialObject = ["price": operation]
index.partialUpdateObject(partialObject, objectID: "myID", block: nil)
```

Note: Here we are incrementing the value by `42`. To increment just by one, put
`value:1`.

Example to decrement a numeric value:

```swift
let operation = [
	"value": 42,
	"_operation": "Decrement"
]
let partialObject = ["price": operation]
index.partialUpdateObject(partialObject, objectID: "myID", block: nil)
```

Note: Here we are decrementing the value by `42`. To decrement just by one, put
`value:1`.

Search
-------------


To perform a search, you only need to initialize the index and perform a call to the search function.

The search query allows only to retrieve 1000 hits, if you need to retrieve more than 1000 hits for seo, you can use [Backup / Retrieve all index content](#backup--retrieve-of-all-index-content)

You can use the following optional arguments:

### Query Parameters

#### Full Text Search Parameters

 * **query**: (string) The instant search query string. All words of the query are interpreted as prefixes (for example "John Mc" will match "John Mccamey" and "Johnathan Mccamey"). If no query parameter is set all objects are retrieved.
 * **queryType**: Selects how the query words are interpreted. It can be one of the following values:
  * **prefixAll**: All query words are interpreted as prefixes.
  * **prefixLast**: Only the last word is interpreted as a prefix (default behavior).
  * **prefixNone**: No query word is interpreted as a prefix. This option is not recommended.
 * **removeWordsIfNoResults**: This option is used to select a strategy in order to avoid having an empty result page. There are three different options:
  * **lastWords**: When a query does not return any results, the last word will be added as optional. The process is repeated with n-1 word, n-2 word, ... until there are results.
  * **firstWords**: When a query does not return any results, the first word will be added as optional. The process is repeated with second word, third word, ... until there are results.
  * **allOptional**: When a query does not return any results, a second trial will be made with all words as optional. This is equivalent to transforming the AND operand between query terms to an OR operand.
  * **none**: No specific processing is done when a query does not return any results (default behavior).
 * **minWordSizefor1Typo**: The minimum number of characters in a query word to accept one typo in this word.<br/>Defaults to 4.
 * **minWordSizefor2Typos**: The minimum number of characters in a query word to accept two typos in this word.<br/>Defaults to 8.
 * **allowTyposOnNumericTokens**: If set to false, it disables typo tolerance on numeric tokens (numbers). Defaults to false.
 * **typoTolerance**: This option allows you to control the number of typos in the result set:

ChangeLog
CHANGELOG

2016-02-05 2.2.1
  * Added support of snippet ellipsis query parameter
	
2015-12-11 2.2.0
  * Added support for optional tag filters

2015-10-20 2.1.0
  * Added support of multiple bounding box for geo-search
  * Added support of polygon for geo-search
  * Added support of automatic radius computation for geo-search
  * Added support of disableTypoToleranceOnAttributes
  * Fix NSJSONSerialization crashes

2015-09-22 2.0.0
  * Upgrade to Swift 2

2015-08-18 1.4.2
  * Fix crash when requesting with a poor connection

2015-07-10 1.4.1
  * Add analyticsTags

2015-07-06 1.4.0
  * Objective-C support
  * New browseFrom methods
  * Add search cache (disabled by default)

2015-06-30 1.3.0
  * Added support of grouping (distinct=3 for example keep the 3 best hits for a distinct key)

2015-06-21 1.2.3
  * add copy constructor for Query

2015-06-16 1.2.2
  * new browse method that takes a Query object

2015-06-09 1.2.1
  * rename `fullTextQuery` to `query`

2015-06-08 1.2.0
  * new retry logic
  * add new parameter on the Query: setMinProximity & setHighlightingTags
  * add new method on the Query: resetLocationParameters


Source/Client.swift
//
//  Copyright (c) 2015 Algolia
//  http://www.algolia.com/
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

import Foundation

#if ALGOLIA_SDK
    import AlgoliaSearchSDK
#endif


/// Entry point in the Swift API.
///
/// You should instantiate a Client object with your AppID, ApiKey and Hosts
/// to start using Algolia Search API.
public class Client : NSObject {
    public var apiKey: String {
        didSet {
            setExtraHeader(apiKey, forKey: "X-Algolia-API-Key")
        }
    }
    
    /// Security tag header
    public var queryParameters: String? {
        didSet {
            if let queryParameters = queryParameters {
                setExtraHeader(queryParameters, forKey: "X-Algolia-QueryParameters")
            } else {
                // FIXME: Should not work (as per the doc).
                manager.session.configuration.HTTPAdditionalHeaders?.removeValueForKey("X-Algolia-QueryParameters")
            }
        }
    }

    /// Security tag header (deprecated)
    public var tagFilters: String? {
        didSet {
            if let tagFilters = tagFilters {
                setExtraHeader(tagFilters, forKey: "X-Algolia-TagFilters")
            } else {
                // FIXME: Should not work (as per the doc).
                manager.session.configuration.HTTPAdditionalHeaders?.removeValueForKey("X-Algolia-TagFilters")
            }
        }
    }

    /// User-token header
    public var userToken: String? {
        didSet {
            if let userToken = userToken {
                setExtraHeader(userToken, forKey: "X-Algolia-UserToken")
            } else {
                // FIXME: Should not work (as per the doc).
                manager.session.configuration.HTTPAdditionalHeaders?.removeValueForKey("X-Algolia-UserToken")
            }
        }
    }

    private let timeout: NSTimeInterval = 30
    private let searchTimeout: NSTimeInterval = 5
    private let incrementTimeout: NSTimeInterval = 10

    public let appID: String

    let readQueryHostnames: [String]
    let writeQueryHostnames: [String]

    let manager: Manager // FIXME: Make private again?
    private var requestBuffer = RingBuffer<Request>(maxCapacity: 10)

    /// Algolia Search initialization.
    ///
    /// - parameter appID: the application ID you have in your admin interface
    /// - parameter apiKey: a valid API key for the service
    /// - parameter queryParameters: value of the header X-Algolia-QueryParameters
    /// - parameter tagFilters: value of the header X-Algolia-TagFilters (deprecated)
    /// - parameter userToken: value of the header X-Algolia-UserToken
    /// - parameter hostnames: the list of hosts that you have received for the service
    public init(appID: String, apiKey: String, queryParameters: String? = nil, tagFilters: String? = nil, userToken: String? = nil, hostnames: [String]? = nil) {
        self.appID = appID
        self.apiKey = apiKey
        self.tagFilters = tagFilters
        self.userToken = userToken
        self.queryParameters = queryParameters

        if let hostnames = hostnames {
            readQueryHostnames = hostnames
            writeQueryHostnames = hostnames
        } else {
            readQueryHostnames = [
                "\(appID)-dsn.algolia.net",
                "\(appID)-1.algolianet.com",
                "\(appID)-2.algolianet.com",
                "\(appID)-3.algolianet.com"
            ]
            
            writeQueryHostnames = [
                "\(appID).algolia.net",
                "\(appID)-1.algolianet.com",
                "\(appID)-2.algolianet.com",
                "\(appID)-3.algolianet.com"
            ]
        }
        
        let version = NSBundle(forClass: self.dynamicType).infoDictionary!["CFBundleShortVersionString"] as! String

        var HTTPHeaders = [
            "X-Algolia-API-Key": self.apiKey,
            "X-Algolia-Application-Id": self.appID,
            "User-Agent": "Algolia for Swift \(version)"
        ]

        if let queryParameters = self.queryParameters {
            HTTPHeaders["X-Algolia-QueryParameters"] = queryParameters
        }
        if let tagFilters = self.tagFilters {
            HTTPHeaders["X-Algolia-TagFilters"] = tagFilters
        }
        if let userToken = self.userToken {
            HTTPHeaders["X-Algolia-UserToken"] = userToken
        }

        manager = Manager(HTTPHeaders: HTTPHeaders)
        
        #if ALGOLIA_SDK
            buildQueue.name = "AlgoliaSearch-Build"
            buildQueue.maxConcurrentOperationCount = 1
            searchQueue.name = "AlgoliaSearch-Search"
            searchQueue.maxConcurrentOperationCount = 1
        #endif
    }
    
    // Helper for Obj-C
    public class func clientWithAppID(appID: String, apiKey: String) -> Client {
        return Client(appID: appID, apiKey: apiKey)
    }
    
    // Helper for Obj-C
    public class func clientWithQueryParameters(queryParameters: String, appID: String, apiKey: String) -> Client {
        return Client(appID: appID, apiKey: apiKey, queryParameters: queryParameters)
    }

    /// Allow to set custom extra header.
    ///
    /// - parameter value: value of the header
    /// - parameter forKey: key of the header
    public func setExtraHeader(value: String, forKey key: String) {
        // FIXME: Should not work (as per the doc).
        if (manager.session.configuration.HTTPAdditionalHeaders != nil) {
            manager.session.configuration.HTTPAdditionalHeaders!.updateValue(value, forKey: key)
        } else {
            let HTTPHeader = [key: value]
            manager.session.configuration.HTTPAdditionalHeaders = HTTPHeader
        }
    }

    // MARK: - Operations

    /// List all existing indexes.
    ///
    /// :return: JSON Object in the handler in the form: { "items": [ {"name": "contacts", "createdAt": "2013-01-18T15:33:13.556Z"}, {"name": "notes", "createdAt": "2013-01-18T15:33:13.556Z"}]}
    public func listIndexes(block: CompletionHandler) {
        performHTTPQuery("1/indexes", method: .GET, body: nil, hostnames: readQueryHostnames, block: block)
    }

    /// Delete an index.
    ///
    /// - parameter indexName: the name of index to delete
    /// :return: JSON Object in the handler containing a "deletedAt" attribute
    public func deleteIndex(indexName: String, block: CompletionHandler? = nil) {
        let path = "1/indexes/\(indexName.urlEncode())"
        performHTTPQuery(path, method: .DELETE, body: nil, hostnames: writeQueryHostnames, block: block)
    }

    /// Move an existing index.
    ///
    /// - parameter srcIndexName: the name of index to move.
    /// - parameter dstIndexName: the new index name that will contains sourceIndexName (destination will be overriten if it already exist).
    public func moveIndex(srcIndexName: String, to dstIndexName: String, block: CompletionHandler? = nil) {
        let path = "1/indexes/\(srcIndexName.urlEncode())/operation"
        let request = [
            "destination": dstIndexName,
            "operation": "move"
        ]

        performHTTPQuery(path, method: .POST, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Copy an existing index.
    ///
    /// - parameter srcIndexName: the name of index to copy.
    /// - parameter dstIndexName: the new index name that will contains a copy of sourceIndexName (destination will be overriten if it already exist).
    public func copyIndex(srcIndexName: String, to dstIndexName: String, block: CompletionHandler? = nil) {
        let path = "1/indexes/\(srcIndexName.urlEncode())/operation"
        let request = [
            "destination": dstIndexName,
            "operation": "copy"
        ]

        performHTTPQuery(path, method: .POST, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Return 10 last log entries.
    public func getLogs(block: CompletionHandler) {
        performHTTPQuery("1/logs", method: .GET, body: nil, hostnames: writeQueryHostnames, block: block)
    }

    /// Return last logs entries.
    ///
    /// - parameter offset: Specify the first entry to retrieve (0-based, 0 is the most recent log entry).
    /// - parameter length: Specify the maximum number of entries to retrieve starting at offset. Maximum allowed value: 1000.
    public func getLogsWithOffset(offset: UInt, length: UInt, block: CompletionHandler) {
        let path = "1/logs?offset=\(offset)&length=\(length)"
        performHTTPQuery(path, method: .GET, body: nil, hostnames: readQueryHostnames, block: block)
    }

    /// Return last logs entries.
    ///
    /// - parameter offset: Specify the first entry to retrieve (0-based, 0 is the most recent log entry).
    /// - parameter length: Specify the maximum number of entries to retrieve starting at offset. Maximum allowed value: 1000.
    public func getLogsWithType(type: String, offset: UInt, length: UInt, block: CompletionHandler) {
        let path = "1/logs?offset=\(offset)&length=\(length)&type=\(type)"
        performHTTPQuery(path, method: .GET, body: nil, hostnames: readQueryHostnames, block: block)
    }

    /// Get the index object initialized (no server call needed for initialization).
    ///
    /// - parameter indexName: the name of index
    public func getIndex(indexName: String) -> Index {
        return Index(client: self, indexName: indexName)
    }

    /// List all existing user keys with their associated ACLs.
    public func listUserKeys(block: CompletionHandler) {
        performHTTPQuery("1/keys", method: .GET, body: nil, hostnames: readQueryHostnames, block: block)
    }

    /// Get ACL of a user key.
    public func getUserKeyACL(key: String, block: CompletionHandler) {
        let path = "1/keys/\(key)"
        performHTTPQuery(path, method: .GET, body: nil, hostnames: readQueryHostnames, block: block)
    }

    /// Delete an existing user key.
    public func deleteUserKey(key: String, block: CompletionHandler? = nil) {
        let path = "1/keys/\(key)"
        performHTTPQuery(path, method: .DELETE, body: nil, hostnames: writeQueryHostnames, block: block)
    }

    /// Create a new user key
    ///
    /// - parameter acls: The list of ACL for this key. The list can contains the following values (as String): search, addObject, deleteObject, deleteIndex, settings, editSettings
    public func addUserKey(acls: [String], block: CompletionHandler? = nil) {
        let request = ["acl": acls]
        performHTTPQuery("1/keys", method: .POST, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Create a new user key
    ///
    /// - parameter acls: The list of ACL for this key. The list can contains the following values (as String): search, addObject, deleteObject, deleteIndex, settings, editSettings
    /// - parameter withValidity: The number of seconds after which the key will be automatically removed (0 means no time limit for this key)
    /// - parameter maxQueriesPerIPPerHour: Specify the maximum number of API calls allowed from an IP address per hour.  Defaults to 0 (unlimited).
    /// - parameter maxHitsPerQuery: Specify the maximum number of hits this API key can retrieve in one call. Defaults to 0 (unlimited)
    public func addUserKey(acls: [String], withValidity validity: UInt, maxQueriesPerIPPerHour maxQueries: UInt, maxHitsPerQuery maxHits: UInt, block: CompletionHandler? = nil) {
        let request: [String: AnyObject] = [
            "acl": acls,
            "validity": validity,
            "maxQueriesPerIPPerHour": maxQueries,
            "maxHitsPerQuery": maxHits,
        ]

        performHTTPQuery("1/keys", method: .POST, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Create a new user key
    ///
    /// - parameter acls: The list of ACL for this key. The list can contains the following values (as String): search, addObject, deleteObject, deleteIndex, settings, editSettings
    /// - parameter forIndexes: restrict this new API key to specific index names
    /// - parameter withValidity: The number of seconds after which the key will be automatically removed (0 means no time limit for this key)
    /// - parameter maxQueriesPerIPPerHour: Specify the maximum number of API calls allowed from an IP address per hour.  Defaults to 0 (unlimited).
    /// - parameter maxHitsPerQuery: Specify the maximum number of hits this API key can retrieve in one call. Defaults to 0 (unlimited)
    public func addUserKey(acls: [String], forIndexes indexes: [String], withValidity validity: UInt, maxQueriesPerIPPerHour maxQueries: UInt, maxHitsPerQuery maxHits: UInt, block: CompletionHandler? = nil) {
        let request: [String: AnyObject] = [
            "acl": acls,
            "indexes": indexes,
            "validity": validity,
            "maxQueriesPerIPPerHour": maxQueries,
            "maxHitsPerQuery": maxHits,
        ]

        performHTTPQuery("1/keys", method: .POST, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Update a user key
    ///
    /// - parameter key: The key to update
    /// - parameter withAcls: The list of ACL for this key. The list can contains the following values (as String): search, addObject, deleteObject, deleteIndex, settings, editSettings
    public func updateUserKey(key: String, withACL acls: [String], block: CompletionHandler? = nil) {
        let path = "1/keys/\(key)"
        let request = ["acl": acls]
        performHTTPQuery(path, method: .PUT, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Update a user key
    ///
    /// - parameter key: The key to update
    /// - parameter withAcls: The list of ACL for this key. The list can contains the following values (as String): search, addObject, deleteObject, deleteIndex, settings, editSettings
    /// - parameter andValidity: The number of seconds after which the key will be automatically removed (0 means no time limit for this key)
    /// - parameter maxQueriesPerIPPerHour: Specify the maximum number of API calls allowed from an IP address per hour.  Defaults to 0 (unlimited).
    /// - parameter maxHitsPerQuery: Specify the maximum number of hits this API key can retrieve in one call. Defaults to 0 (unlimited)
    public func updateUserKey(key: String, withACL acls: [String], andValidity validity: UInt, maxQueriesPerIPPerHour maxQueries: UInt, maxHitsPerQuery maxHits: UInt, block: CompletionHandler? = nil) {
        let path = "1/keys/\(key)"
        let request: [String: AnyObject] = [
            "acl": acls,
            "validity": validity,
            "maxQueriesPerIPPerHour": maxQueries,
            "maxHitsPerQuery": maxHits,
        ]

        performHTTPQuery(path, method: .PUT, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Update a user key
    ///
    /// - parameter key: The key to update
    /// - parameter withAcls: The list of ACL for this key. The list can contains the following values (as String): search, addObject, deleteObject, deleteIndex, settings, editSettings
    /// - parameter andValidity: The number of seconds after which the key will be automatically removed (0 means no time limit for this key)
    /// - parameter forIndexes: restrict this API key to specific index names
    /// - parameter maxQueriesPerIPPerHour: Specify the maximum number of API calls allowed from an IP address per hour.  Defaults to 0 (unlimited).
    /// - parameter maxHitsPerQuery: Specify the maximum number of hits this API key can retrieve in one call. Defaults to 0 (unlimited)
    public func updateUserKey(key: String, withACL acls: [String], andValidity validity: UInt, forIndexes indexes: [String], maxQueriesPerIPPerHour maxQueries: UInt, maxHitsPerQuery maxHits: UInt, block: CompletionHandler? = nil) {
        let path = "1/keys/\(key)"
        let request: [String: AnyObject] = [
            "acl": acls,
            "indexes": indexes,
            "validity": validity,
            "maxQueriesPerIPPerHour": maxQueries,
            "maxHitsPerQuery": maxHits,
        ]

        performHTTPQuery(path, method: .PUT, body: request, hostnames: writeQueryHostnames, block: block)
    }

    /// Query multiple indexes with one API call.
    ///
    /// - parameter queries: An array of queries with the associated index (Array of Dictionnary object ["indexName": "targettedIndex", "query": QueryObject]).
    public func multipleQueries(queries: [AnyObject], block: CompletionHandler? = nil) {
        let path = "1/indexes/*/queries"

        var convertedQueries = [[String: String]]()
        convertedQueries.reserveCapacity(queries.count)
        for query in queries {
            if let query = query as? [String: AnyObject] {
                convertedQueries.append([
                    "params": (query["query"] as! Query).buildURL(),
                    "indexName": query["indexName"] as! String
                    ])
            }
        }

        let request = ["requests": convertedQueries]
        performHTTPQuery(path, method: .POST, body: request, hostnames: readQueryHostnames, block: block)
    }

    // MARK: - Network

    /// Perform an HTTP Query.
    func performHTTPQuery(path: String, method: HTTPMethod, body: [String: AnyObject]?, hostnames: [String], isSearchQuery: Bool = false, index: Int = 0, block: CompletionHandler? = nil) {
        assert(index < hostnames.count, "\(index) < \(hostnames.count) !")

        var currentTimeout = (isSearchQuery) ? searchTimeout : timeout
        if index > 1 {
            currentTimeout += incrementTimeout
        }
        // FIXME: This should not work... at least on iOS >= 9. As per the doc:
        // > Changing mutable values within the configuration object has no effect on the current session, but you can
        // > create a new session with the modified configuration object.
        manager.session.configuration.timeoutIntervalForRequest = currentTimeout

        let request = manager.request(method, "https://\(hostnames[index])/\(path)", parameters: body) { (response, data, error) -> Void in
            if let statusCode = response?.statusCode {
                if let block = block {
                    switch(statusCode) {
                    case 200..<300 where data is [String: AnyObject]:
                        block(content: (data as! [String: AnyObject]), error: nil)
                    default:
                        if let data = data as? [String: AnyObject], errorMessage = data["message"] as? String {
                            block(content: nil, error: NSError(domain: AlgoliaSearchErrorDomain, code: statusCode, userInfo: [NSLocalizedDescriptionKey: errorMessage]))
                        } else {
                            block(content: nil, error: NSError(domain: AlgoliaSearchErrorDomain, code: statusCode, userInfo: nil))
                        }
                    }
                }
            } else {
                if (index + 1) < hostnames.count {
                    self.performHTTPQuery(path, method: method, body: body, hostnames: hostnames, isSearchQuery: isSearchQuery, index: index + 1, block: block)
                } else {
                    block?(content: nil, error: error)
                }
            }
        }

        requestBuffer.append(request)
    }

    /// Cancel a queries. Only the last ten queries can be cancelled.
    func cancelQueries(method: HTTPMethod, path: String) {
        for request in requestBuffer {
            if request.request.URL!.path! == path {
                if request.request.HTTPMethod == method.rawValue {
                    request.cancel()
                }
            }
        }
    }
    
// ----------------------------------------------------------------------
// NOTICE: Start of SDK-dependent code
// ----------------------------------------------------------------------

#if ALGOLIA_SDK
    var sdk: ASSdk = ASSdk.sharedSdk()
    var rootDataDir: String?

    /// Queue used to build local indices in the background.
    let buildQueue = NSOperationQueue()
    /// Queue used to search local indices in the background.
    let searchQueue = NSOperationQueue()

    public func enableOfflineMode(licenseData: NSData) {
        // Create the cache directory.
        do {
            self.rootDataDir = NSSearchPathForDirectoriesInDomains(.CachesDirectory, .UserDomainMask, true).first! + "/algolia"
            try NSFileManager.defaultManager().createDirectoryAtPath(self.rootDataDir!, withIntermediateDirectories: true, attributes: nil)
        } catch _ {
            NSLog("Error: could not create cache directory '%@'", self.rootDataDir!)
        }
        
        // Init the SDK.
        sdk.initWithLicenseData(licenseData)
    }
    
#endif // ALGOLIA_SDK

// ----------------------------------------------------------------------
// NOTICE: End of SDK-dependent code
// ----------------------------------------------------------------------

}
